sql/tests: TestRandomSyntaxSQLSmith failed

sql/tests.TestRandomSyntaxSQLSmith [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3473202&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3473202&tab=artifacts#/) on master @ [676dc076dc6a137d88eac9217764bbb63beae86f](https://github.com/cockroachdb/cockroach/commits/676dc076dc6a137d88eac9217764bbb63beae86f):

Random syntax error:

```
    rsg_test.go:779: Crash detected: server panic: pq: internal error: unsupported comparison: oid to bytes
```
Query:

```
        SELECT
        	(-1):::INT8 AS col_1128285,
        	tab_532869.col2_1 AS col_1128286,
        	array_position(ARRAY[(1173684112:::OID,),(NULL,)], ('\xc03b4478eb':::BYTES,))::INT8 AS col_1128287,
        	tab_532867.crdb_internal_idx_expr AS col_1128288,
        	tab_532870.crdb_internal_mvcc_timestamp AS col_1128289,
        	NULL AS col_1128290,
        	tab_532868.tableoid AS col_1128291,
        	tab_532868.tableoid AS col_1128292,
        	'1992-11-20 03:49:26.000695+00:00':::TIMESTAMPTZ AS col_1128293,
        	(tab_532867.col2_3::INT8 >> (-3486767938233063631):::INT8::INT8)::INT8 AS col_1128294,
        	(-1):::INT8 AS col_1128295
        FROM
        	defaultdb.public.table2@table2_col2_0_col2_1_col2_2_col2_3_idx AS tab_532867,
        	defaultdb.public.table2@[0] AS tab_532868
        	JOIN defaultdb.public.table2@[0] AS tab_532869 ON false,
        	defaultdb.public.table2@table2_col2_2_expr_col2_0_expr1_idx AS tab_532870
        WHERE
        	true;
```
Schema:

```
    rsg_test.go:596: To reproduce, use schema:
    rsg_test.go:598: 
        		SET CLUSTER SETTING sql.stats.automatic_collection.enabled = false;
        		SET CLUSTER SETTING sql.stats.histogram_collection.enabled = false;
        		SET CLUSTER SETTING sql.defaults.interleaved_tables.enabled = true;
        	CREATE TABLE table1 (col1_0 INT2 NULL, INDEX (col1_0), UNIQUE (col1_0 DESC) WHERE table1.col1_0 <= 32767:::INT8, INDEX (col1_0), INDEX (col1_0 ASC), UNIQUE (col1_0), INDEX (col1_0) WHERE table1.col1_0 > (-1):::INT8, INDEX (col1_0 DESC) WHERE table1.col1_0 >= 32767:::INT8, INDEX (col1_0), UNIQUE (col1_0));
        CREATE TABLE table2 (col2_0 TIMESTAMPTZ NOT NULL, col2_1 REGPROCEDURE NOT NULL, col2_2 INT4 NOT NULL, col2_3 INT8 NOT NULL, PRIMARY KEY (col2_1 DESC, col2_0 ASC, col2_3), INDEX (col2_2 DESC, col2_0 DESC), UNIQUE (col2_3 DESC, col2_0 ASC) WHERE table2.col2_3 != 32767:::INT8, INDEX (col2_0 DESC, col2_1 ASC, col2_2, col2_3 DESC) WHERE ((table2.col2_2 >= 0:::INT8) OR (table2.col2_0 != '294276-12-31 23:59:59.999999+00:00':::TIMESTAMPTZ)) OR (table2.col2_3 > (-128):::INT8), UNIQUE (col2_1 ASC), INDEX (col2_2 ASC, (CASE WHEN col2_0 IS NULL THEN e'X\x18$\x11I6\r':::STRING ELSE e'C\b\'/6(':::STRING END), col2_0 ASC, (col2_2 + col2_3) DESC), UNIQUE (col2_1, col2_0 DESC, col2_2 ASC) WHERE ((table2.col2_0 = '3000-01-01 00:00:00+00:00':::TIMESTAMPTZ) OR (table2.col2_3 >= (-2147483648):::INT8)) OR (table2.col2_2 >= 1:::INT8), FAMILY (col2_0), FAMILY (col2_2, col2_3, col2_1), UNIQUE (col2_2));
        ALTER TABLE table1 INJECT STATISTICS '[{"columns": ["col1_0"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 5104017871428977943, "histo_buckets": [{"distinct_range": 0, "num_eq": 40000000, "num_range": 0, "upper_bound": "13093"}], "histo_col_type": "INT2", "name": "__auto__", "null_count": 8989897701368440168, "row_count": 9055058578208466492}]':::JSONB;
        ALTER TABLE table2 INJECT STATISTICS '[{"columns": ["col2_2"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 7904702, "histo_buckets": [{"distinct_range": 0, "num_eq": 6921569931402439194, "num_range": 0, "upper_bound": "-1675587582"}, {"distinct_range": 1620191154237186600, "num_eq": 101348852754977614, "num_range": 8976110562440874917, "upper_bound": "-572419750"}, {"distinct_range": 404.33391430455134, "num_eq": 500000000, "num_range": 500, "upper_bound": "1147744038"}], "histo_col_type": "INT4", "name": "__auto__", "null_count": 0, "row_count": 8000000}, {"columns": ["col2_3"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 4845613, "histo_buckets": [{"distinct_range": 0, "num_eq": 8000000000, "num_range": 0, "upper_bound": "-8876981534753377004"}], "histo_col_type": "INT8", "name": "__auto__", "null_count": 0, "row_count": 8000000}, {"columns": ["col2_0"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 5987192, "histo_buckets": [{"distinct_range": 0, "num_eq": 60000, "num_range": 0, "upper_bound": "2029-03-27 21:30:43.000333+00:00"}], "histo_col_type": "TIMESTAMPTZ", "name": "__auto__", "null_count": 0, "row_count": 8000000}, {"columns": ["col2_1"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 3637731, "histo_buckets": [{"distinct_range": 0, "num_eq": 50, "num_range": 0, "upper_bound": "1638559855"}, {"distinct_range": 0, "num_eq": 50000000000, "num_range": 0, "upper_bound": "2729383840"}, {"distinct_range": 70000000, "num_eq": 2627280684520698847, "num_range": 70000000, "upper_bound": "3647080921"}, {"distinct_range": 342.67100677751387, "num_eq": 3000, "num_range": 2000, "upper_bound": "3654389876"}, {"distinct_range": 1037395906969251200, "num_eq": 3200707335841305569, "num_range": 1037395906969251154, "upper_bound": "4006606552"}], "histo_col_type": "REGPROCEDURE", "name": "__auto__", "null_count": 0, "row_count": 8000000}]':::JSONB;
        ALTER TABLE table1 ADD FOREIGN KEY (col1_0) REFERENCES table2 (col2_2) ON DELETE SET DEFAULT;
        CREATE TYPE rand_typ_0 AS ENUM ('rmyhik', 'uxb');
        CREATE TYPE rand_typ_1 AS ENUM ('wo');
        CREATE TYPE rand_typ_2 AS ENUM ('nye', 'a');
        ;
    rsg_test.go:598: 
        CREATE TYPE greeting AS ENUM ('hello', 'howdy', 'hi', 'good day', 'morning');
        CREATE TABLE IF NOT EXISTS seed AS
        	SELECT
        		g::INT2 AS _int2,
        		g::INT4 AS _int4,
        		g::INT8 AS _int8,
        		g::FLOAT4 AS _float4,
        		g::FLOAT8 AS _float8,
        		'2001-01-01'::DATE + g AS _date,
        		'2001-01-01'::TIMESTAMP + g * '1 day'::INTERVAL AS _timestamp,
        		'2001-01-01'::TIMESTAMPTZ + g * '1 day'::INTERVAL AS _timestamptz,
        		g * '1 day'::INTERVAL AS _interval,
        		g % 2 = 1 AS _bool,
        		g::DECIMAL AS _decimal,
        		g::STRING AS _string,
        		g::STRING::BYTES AS _bytes,
        		substring('00000000-0000-0000-0000-' || g::STRING || '00000000000', 1, 36)::UUID AS _uuid,
        		'0.0.0.0'::INET + g AS _inet,
        		g::STRING::JSONB AS _jsonb,
        		enum_range('hello'::greeting)[g] as _enum
        	FROM
        		generate_series(1, 5) AS g;
        
        INSERT INTO seed DEFAULT VALUES;
        CREATE INDEX on seed (_int8, _float8, _date);
        CREATE INVERTED INDEX on seed (_jsonb);
        ;
    rsg_test.go:600: 
    rsg_test.go:601: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestRandomSyntaxSQLSmith274670283
--- FAIL: TestRandomSyntaxSQLSmith (300.97s)
```
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestRandomSyntaxSQLSmith PKG=./pkg/sql/tests TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #70000 sql/tests: TestRandomSyntaxSQLSmith failed [NPE in index join] [C-test-failure GA-blocker O-robot T-sql-queries branch-release-21.2]
- #64592 sql/tests: TestRandomSyntaxSQLSmith failed [C-test-failure O-robot T-sql-queries branch-release-21.1]
</p>
</details>
<details><summary>Internal log</summary>
<p>

```
mjibson marked as alumn{us/a}; resolving to rafiss instead
```
</p>
</details>

/cc @cockroachdb/sql-experience rafiss
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRandomSyntaxSQLSmith.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql/tests.TestRandomSyntaxSQLSmith [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3479919&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3479919&tab=artifacts#/) on master @ [1dbea6a3e4c93ba7c4a43ad34369c438558eba8a](https://github.com/cockroachdb/cockroach/commits/1dbea6a3e4c93ba7c4a43ad34369c438558eba8a):

Random syntax error:

```
    rsg_test.go:779: Crash detected: server panic: pq: internal error: interface conversion: tree.Datum is *tree.DString, not *tree.DGeography
```
Query:

```
        WITH
        	with_57996 (col_329245)
        		AS (
        			SELECT
        				*
        			FROM
        				(
        					VALUES
        						(
        							array_position(ARRAY[('01020000A0E610000003000000303E5E18C5C24A4078F5D198967A2E406A78D5DBFC1AF6C1000F8BE1A2450F408056D975D63106C0DE39E8E4653D02C274007C9C5B335CC060EE05D22CD54E4040BE88F65068E641':::GEOGRAPHY,)], ('G':::STRING,))::INT8
        						),
        						((-2698298997321294264):::INT8),
        						(4126820061119538877:::INT8),
        						((-5242058052621529232):::INT8),
        						((-8494155643475552972):::INT8)
        				)
        					AS tab_141968 (col_329245)
        		)
        SELECT
        	cte_ref_16700.col_329245 AS col_329246
        FROM
        	with_57996 AS cte_ref_16700
        WHERE
        	NULL
        ORDER BY
        	cte_ref_16700.col_329245 DESC,
        	cte_ref_16700.col_329245 ASC,
        	cte_ref_16700.col_329245 DESC,
        	cte_ref_16700.col_329245,
        	cte_ref_16700.col_329245;
```
Schema:

```
    rsg_test.go:596: To reproduce, use schema:
    rsg_test.go:598: 
        		SET CLUSTER SETTING sql.stats.automatic_collection.enabled = false;
        		SET CLUSTER SETTING sql.stats.histogram_collection.enabled = false;
        		SET CLUSTER SETTING sql.defaults.interleaved_tables.enabled = true;
        	CREATE TABLE table1 (col1_0 BOX2D NOT NULL, col1_1 REGCLASS NULL, col1_2 VARCHAR NOT NULL, col1_3 FLOAT4 NULL, col1_4 BOOL NULL, col1_5 GEOMETRY NOT NULL, col1_6 UUID NULL, col1_7 INTERVAL NOT NULL, col1_8 INT2 NOT NULL, col1_9 REGPROC NULL, col1_10 INT8 NOT NULL, col1_11 DATE NOT NULL, col1_12 INT4, col1_13 INET, col1_14 TIME NULL, col1_15 INT2 NOT NULL, col1_16 REGNAMESPACE NULL, col1_17 INT4 NOT NULL, col1_18 STRING AS (lower(CAST(col1_13 AS STRING))) STORED, UNIQUE (col1_4 DESC, col1_10 ASC, lower(CAST(col1_9 AS STRING)) ASC, col1_1 DESC, col1_14 ASC, col1_15, col1_9 DESC, col1_12 ASC, col1_13 ASC, col1_11, col1_6 ASC, col1_0), UNIQUE (col1_16 ASC, col1_10 ASC, col1_14 ASC, col1_9 ASC, col1_6 DESC, col1_2 ASC, col1_11 DESC, col1_7, col1_18 DESC, col1_8, col1_17, col1_13) STORING (col1_1, col1_4, col1_5, col1_12, col1_15), UNIQUE (col1_15 ASC, col1_13, ((col1_8 + col1_17) + col1_12), col1_0 ASC, lower(CAST(col1_9 AS STRING)) ASC, col1_12 DESC, col1_9 DESC, (col1_15 + col1_8) ASC, col1_18 DESC, col1_1 ASC, col1_11 DESC, col1_7, col1_4 ASC, col1_17 ASC, col1_2 DESC, col1_10) STORING (col1_5, col1_6), INDEX (col1_8 ASC, col1_14, col1_1 ASC, col1_11 DESC) STORING (col1_5, col1_6, col1_9, col1_13, col1_15, col1_17), UNIQUE (col1_6));
        CREATE TABLE table2 (col2_0 OID NOT NULL, col2_1 NAME NOT NULL, col2_2 TIMESTAMP NOT NULL, col2_3 INTERVAL NOT NULL, col2_4 CHAR NOT NULL, PRIMARY KEY (col2_2 DESC, col2_1 DESC, col2_0 ASC, col2_4, col2_3 DESC), UNIQUE (col2_4 DESC), INDEX (lower(col2_4) DESC, col2_0), UNIQUE (col2_1 DESC), INDEX (col2_4 ASC, col2_2 ASC, col2_3 ASC), INDEX (col2_2 DESC), FAMILY (col2_3, col2_2), FAMILY (col2_4, col2_0, col2_1));
        CREATE TABLE table3 (col3_0 REGROLE NULL, col3_1 BOOL NOT NULL, col3_2 GEOMETRY, col3_3 DECIMAL, col3_4 STRING NULL AS (lower(CAST(col3_0 AS STRING))) STORED, col3_5 DECIMAL AS (col3_3 + (-1940483658809.848209):::DECIMAL) VIRTUAL, col3_6 DECIMAL AS (col3_3 + 1.868796235169743972E+27:::DECIMAL) VIRTUAL, FAMILY (col3_0), FAMILY (col3_1, col3_4), FAMILY (col3_3), FAMILY (col3_2));
        CREATE TABLE table4 (col4_0 TIME, col4_1 INET NOT NULL, col4_2 REGPROCEDURE NOT NULL, col4_3 VARCHAR NOT NULL, col4_4 BOX2D[] NOT NULL, col4_5 STRING AS (lower(CAST(col4_0 AS STRING))) VIRTUAL, UNIQUE (col4_1 ASC, lower(CAST(col4_1 AS STRING)) ASC, col4_0 ASC, col4_3, col4_5 ASC, lower(CAST(col4_1 AS STRING)) DESC) STORING (col4_4) WHERE ((table4.col4_3 > e'\x00':::STRING) AND (table4.col4_0 < '00:00:00':::TIME)) AND (table4.col4_5 = e'\x00':::STRING), INDEX (col4_5 DESC, col4_0 DESC, col4_2 ASC, col4_3 DESC) WHERE ((table4.col4_3 >= e'\U00002603':::STRING) AND (table4.col4_5 <= '':::STRING)) OR (table4.col4_0 >= '00:00:00':::TIME), INDEX (col4_0));
        ALTER TABLE table1 INJECT STATISTICS '[{"columns": ["col1_1"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 6410258510711274195, "histo_col_type": "", "name": "__auto__", "null_count": 37624480388708407, "row_count": 7757861913891637596}, {"columns": ["col1_2"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 6418447343184041220, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 7757861913891637596}, {"columns": ["col1_10"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 5383916556226412315, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 7757861913891637596}, {"columns": ["col1_3"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 3309879125904428775, "histo_col_type": "", "name": "__auto__", "null_count": 3186029012837096609, "row_count": 7757861913891637596}, {"columns": ["col1_4"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 3067334272420961211, "histo_buckets": [{"distinct_range": 0, "num_eq": 0, "num_range": 0, "upper_bound": "false"}, {"distinct_range": 2471563610814379500, "num_eq": 6000000, "num_range": 5089647440644737151, "upper_bound": "true"}], "histo_col_type": "BOOL", "name": "__auto__", "null_count": 7448651276164749029, "row_count": 7757861913891637596}, {"columns": ["col1_11"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 2535014211556160328, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 7757861913891637596}, {"columns": ["col1_15"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 7640487818880702824, "histo_buckets": [{"distinct_range": 0, "num_eq": 700000000, "num_range": 0, "upper_bound": "-27796"}, {"distinct_range": 0, "num_eq": 7338726663878710724, "num_range": 400000000, "upper_bound": "1175"}, {"distinct_range": 2412267947142285000, "num_eq": 50000, "num_range": 3849945528713182451, "upper_bound": "15358"}, {"distinct_range": 437758869.006742, "num_eq": 111076191220369820, "num_range": 1000000000, "upper_bound": "17004"}, {"distinct_range": 2848305046033236500, "num_eq": 5131673637443665483, "num_range": 3753941202882479265, "upper_bound": "22923"}], "histo_col_type": "INT2", "name": "__auto__", "null_count": 0, "row_count": 7757861913891637596}, {"columns": ["col1_17"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 3583404929116252944, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 7757861913891637596}, {"columns": ["col1_18"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 5331860191878561375, "histo_col_type": "", "name": "__auto__", "null_count": 2521262619086793063, "row_count": 7757861913891637596}, {"columns": ["col1_0"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 4058539729842313402, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 7757861913891637596}, {"columns": ["col1_6"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 2857204368789096841, "histo_buckets": [{"distinct_range": 0, "num_eq": 20000000, "num_range": 0, "upper_bound": "00000000-0000-0000-0000-000000000000"}, {"distinct_range": 1167152931179623200, "num_eq": 6000000000, "num_range": 1462125467249710739, "upper_bound": "1e6d25c8-2d1a-4110-8f44-5b8fb57d4f33"}, {"distinct_range": 0, "num_eq": 6509605852977458053, "num_range": 10000000000, "upper_bound": "1e85dd40-4745-4784-8b8f-7b46903bfb6c"}, {"distinct_range": 0, "num_eq": 20000, "num_range": 1349621350182318350, "upper_bound": "47c77eca-e706-4705-9bfa-67255da1e58f"}, {"distinct_range": 0, "num_eq": 4754917538398387255, "num_range": 100000, "upper_bound": "521fe25f-c2f1-4ef6-81ae-b502ad7baa58"}, {"distinct_range": 5891032144134723000, "num_eq": 5104174767887400561, "num_range": 5891032144134722255, "upper_bound": "aef6a22b-d11c-4833-9cae-153723e6d36f"}, {"distinct_range": 4137539796835293700, "num_eq": 1000, "num_range": 4137539796835293844, "upper_bound": "bdace3f1-1c5f-4e4e-83f3-68272a905358"}, {"distinct_range": 299.27798135771997, "num_eq": 100000, "num_range": 300, "upper_bound": "e6e459c9-c821-4cdf-8979-8b33a5ed9fd6"}], "histo_col_type": "UUID", "name": "__auto__", "null_count": 6838054370286735716, "row_count": 7757861913891637596}, {"columns": ["col1_7"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 1111475612391053356, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 7757861913891637596}, {"columns": ["col1_13"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 6352370219848080592, "histo_col_type": "", "name": "__auto__", "null_count": 80603139892426380, "row_count": 7757861913891637596}, {"columns": ["col1_14"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 5997756123118448505, "histo_col_type": "", "name": "__auto__", "null_count": 6062999999127516788, "row_count": 7757861913891637596}, {"columns": ["col1_5"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 3684905256352930403, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 7757861913891637596}, {"columns": ["col1_8"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 1576278675174637842, "histo_buckets": [{"distinct_range": 0, "num_eq": 4447667347874315655, "num_range": 0, "upper_bound": "10923"}, {"distinct_range": 0, "num_eq": 6319057645052327, "num_range": 0, "upper_bound": "30049"}], "histo_col_type": "INT2", "name": "__auto__", "null_count": 0, "row_count": 7757861913891637596}, {"columns": ["col1_9"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 2417885713964949573, "histo_col_type": "", "name": "__auto__", "null_count": 7208791624213613960, "row_count": 7757861913891637596}, {"columns": ["col1_12"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 2965797750646387672, "histo_col_type": "", "name": "__auto__", "null_count": 287117231679340338, "row_count": 7757861913891637596}, {"columns": ["col1_16"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 3088318322257652304, "histo_col_type": "REGNAMESPACE", "name": "__auto__", "null_count": 222735264152035241, "row_count": 7757861913891637596}]':::JSONB;
        ALTER TABLE table2 INJECT STATISTICS e'[{"columns": ["col2_0"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 372, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 8000}, {"columns": ["col2_1"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 2309, "histo_buckets": [{"distinct_range": 0, "num_eq": 4854218129624908356, "num_range": 0, "upper_bound": "E\\u0011(BQ"}, {"distinct_range": 161731884.31218195, "num_eq": 90000000000, "num_range": 300000000, "upper_bound": "\\\\xm\\u0004\\"&"}], "histo_col_type": "NAME", "name": "__auto__", "null_count": 0, "row_count": 8000}, {"columns": ["col2_2"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 2836, "histo_buckets": [{"distinct_range": 0, "num_eq": 2964692733964495806, "num_range": 0, "upper_bound": "-4713-11-24 00:00:00"}, {"distinct_range": 1584883907949051100, "num_eq": 3000000000, "num_range": 7621494684790287791, "upper_bound": "1972-03-24 22:14:31.000245"}, {"distinct_range": 0, "num_eq": 10000000000, "num_range": 8000000000, "upper_bound": "1980-03-26 18:20:56.000758"}, {"distinct_range": 5441330.2431649165, "num_eq": 1322065427767895485, "num_range": 9000000, "upper_bound": "2014-06-27 09:23:05.000542"}, {"distinct_range": 194425574381599970, "num_eq": 5000000000, "num_range": 341791324912628894, "upper_bound": "2022-12-26 07:50:51.00013"}, {"distinct_range": 2148999170788176000, "num_eq": 9062375424223278132, "num_range": 2148999170788175791, "upper_bound": "2030-10-08 10:44:16.000727"}, {"distinct_range": 0, "num_eq": 90000, "num_range": 300, "upper_bound": "2033-01-20 16:39:40.000574"}], "histo_col_type": "TIMESTAMP", "name": "__auto__", "null_count": 0, "row_count": 8000}, {"columns": ["col2_3"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 2758, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 8000}, {"columns": ["col2_4"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 1765, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 8000}]':::JSONB;
        ALTER TABLE table3 INJECT STATISTICS '[{"columns": ["col3_0"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 20, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 30}, {"columns": ["col3_1"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 12, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 30}, {"columns": ["col3_2"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 12, "histo_col_type": "", "name": "__auto__", "null_count": 19, "row_count": 30}, {"columns": ["col3_3"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 27, "histo_col_type": "", "name": "__auto__", "null_count": 5, "row_count": 30}, {"columns": ["col3_4"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 23, "histo_col_type": "", "name": "__auto__", "null_count": 6, "row_count": 30}, {"columns": ["col3_5"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 0, "histo_col_type": "", "name": "__auto__", "null_count": 26, "row_count": 30}, {"columns": ["col3_6"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 20, "histo_col_type": "", "name": "__auto__", "null_count": 20, "row_count": 30}]':::JSONB;
        ALTER TABLE table4 INJECT STATISTICS e'[{"columns": ["col4_5"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 3545841576814979442, "histo_buckets": [{"distinct_range": 0, "num_eq": 1000000000, "num_range": 0, "upper_bound": ""}, {"distinct_range": 0, "num_eq": 73973811167108350, "num_range": 0, "upper_bound": "\\u0007z\\u000e\\u001a*\\u001b7G>"}, {"distinct_range": 5000000, "num_eq": 900000000, "num_range": 5000000, "upper_bound": "\\u001d<``"}, {"distinct_range": 1584669951409704400, "num_eq": 700, "num_range": 6761164315376646864, "upper_bound": ",I"}], "histo_col_type": "STRING", "name": "__auto__", "null_count": 838029852614591584, "row_count": 8674633523651056601}, {"columns": ["col4_0"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 4274424613768943059, "histo_buckets": [{"distinct_range": 0, "num_eq": 70, "num_range": 0, "upper_bound": "02:24:03.060203"}, {"distinct_range": 0, "num_eq": 40000000000, "num_range": 700000, "upper_bound": "21:39:02.413006"}], "histo_col_type": "TIME", "name": "__auto__", "null_count": 3597801595363304303, "row_count": 8674633523651056601}, {"columns": ["col4_1"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 3703499219189495780, "histo_buckets": [{"distinct_range": 0, "num_eq": 6647687574370486336, "num_range": 0, "upper_bound": "ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff"}], "histo_col_type": "INET", "name": "__auto__", "null_count": 0, "row_count": 8674633523651056601}, {"columns": ["col4_2"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 8019985163825029526, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 8674633523651056601}, {"columns": ["col4_3"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 3283990849439093242, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 8674633523651056601}, {"columns": ["col4_4"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 6979863265706060477, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 8674633523651056601}]':::JSONB;
        CREATE TYPE rand_typ_0 AS ENUM ('s', 'ikb', 'b', 'ws', 'arho');
        ;
    rsg_test.go:598: 
        CREATE TYPE greeting AS ENUM ('hello', 'howdy', 'hi', 'good day', 'morning');
        CREATE TABLE IF NOT EXISTS seed AS
        	SELECT
        		g::INT2 AS _int2,
        		g::INT4 AS _int4,
        		g::INT8 AS _int8,
        		g::FLOAT4 AS _float4,
        		g::FLOAT8 AS _float8,
        		'2001-01-01'::DATE + g AS _date,
        		'2001-01-01'::TIMESTAMP + g * '1 day'::INTERVAL AS _timestamp,
        		'2001-01-01'::TIMESTAMPTZ + g * '1 day'::INTERVAL AS _timestamptz,
        		g * '1 day'::INTERVAL AS _interval,
        		g % 2 = 1 AS _bool,
        		g::DECIMAL AS _decimal,
        		g::STRING AS _string,
        		g::STRING::BYTES AS _bytes,
        		substring('00000000-0000-0000-0000-' || g::STRING || '00000000000', 1, 36)::UUID AS _uuid,
        		'0.0.0.0'::INET + g AS _inet,
        		g::STRING::JSONB AS _jsonb,
        		enum_range('hello'::greeting)[g] as _enum
        	FROM
        		generate_series(1, 5) AS g;
        
        INSERT INTO seed DEFAULT VALUES;
        CREATE INDEX on seed (_int8, _float8, _date);
        CREATE INVERTED INDEX on seed (_jsonb);
        ;
    rsg_test.go:600: 
    rsg_test.go:601: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestRandomSyntaxSQLSmith661084232
--- FAIL: TestRandomSyntaxSQLSmith (345.83s)
```
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestRandomSyntaxSQLSmith PKG=./pkg/sql/tests TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #64592 sql/tests: TestRandomSyntaxSQLSmith failed [C-test-failure O-robot T-sql-queries branch-release-21.1]
</p>
</details>
<details><summary>Internal log</summary>
<p>

```
mjibson marked as alumn{us/a}; resolving to rafiss instead
```
</p>
</details>

/cc @cockroachdb/sql-experience rafiss
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRandomSyntaxSQLSmith.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql/tests.TestRandomSyntaxSQLSmith [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3485083&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3485083&tab=artifacts#/) on master @ [a465b0d5d51a6534d39e0683fd037d132bcbd576](https://github.com/cockroachdb/cockroach/commits/a465b0d5d51a6534d39e0683fd037d132bcbd576):

Random syntax error:

```
    rsg_test.go:779: Crash detected: server panic: pq: internal error: unsupported comparison: string to oid[]
```
Query:

```
        SELECT
        	e'\x0e\x1f\b?\x16S0':::STRING AS col_17066,
        	tab_6902.col1_1 AS col_17067,
        	'':::STRING AS col_17068,
        	(-8430391825423080272):::DECIMAL AS col_17069,
        	(-5760015391639279772):::INT8 AS col_17070,
        	B'01100101101000001001011010100111111111111110100' AS col_17071,
        	array_position(ARRAY[(e'8\x1dCI':::STRING,),(NULL,)], (ARRAY[1798497560:::OID,3562717310:::OID],))::INT8
        		AS col_17072,
        	tab_6902.col1_8 AS col_17073,
        	tab_6902.col1_6 AS col_17074,
        	5889424387893209298:::INT8 AS col_17075
        FROM
        	defaultdb.public.table1@[0] AS tab_6902
        ORDER BY
        	tab_6902.crdb_internal_idx_expr_1 ASC, tab_6902.col1_6
        LIMIT
        	43:::INT8;
```
Schema:

```
    rsg_test.go:596: To reproduce, use schema:
    rsg_test.go:598: 
        		SET CLUSTER SETTING sql.stats.automatic_collection.enabled = false;
        		SET CLUSTER SETTING sql.stats.histogram_collection.enabled = false;
        		SET CLUSTER SETTING sql.defaults.interleaved_tables.enabled = true;
        	CREATE TABLE table1 (col1_0 BIT(47) NOT NULL, col1_1 TIMESTAMP NOT NULL, col1_2 VARCHAR NOT NULL, col1_3 STRING NOT NULL, col1_4 CHAR NOT NULL, col1_5 INT8, PRIMARY KEY (col1_1 DESC, col1_4 DESC, col1_0 ASC, col1_2 DESC, col1_3 ASC), col1_6 STRING NOT NULL AS (lower(CAST(col1_0 AS STRING))) STORED, col1_7 STRING NOT NULL AS (lower(col1_4)) VIRTUAL, col1_8 INT8 AS (col1_5 + (-3274287437767374164):::INT8) VIRTUAL, col1_9 INT8 AS (col1_5 + NULL) STORED, UNIQUE (col1_8, col1_5, col1_4, col1_6, lower(CAST(col1_1 AS STRING)), col1_0 ASC, col1_2 ASC, col1_3 DESC, col1_9 ASC) WHERE (((((((table1.col1_5 != (-1):::INT8) OR (table1.col1_7 > '"':::STRING)) AND (table1.col1_4 = '"':::STRING)) OR (table1.col1_6 > '':::STRING)) OR (table1.col1_9 <= (-1):::INT8)) AND (table1.col1_2 = 'X':::STRING)) AND (table1.col1_1 >= '-2000-01-01 00:00:00':::TIMESTAMP)) AND (table1.col1_3 = e'\U00002603':::STRING), INDEX (col1_0 DESC, col1_1), UNIQUE (col1_9 DESC, col1_8, col1_2 ASC, col1_7, (col1_5 + 560572723049112575:::INT8) DESC, col1_1 DESC, lower(CAST(col1_1 AS STRING)) DESC) STORING (col1_5, col1_6), UNIQUE (col1_2, col1_7 DESC, lower(CAST(col1_0 AS STRING)) DESC, col1_8, col1_3 DESC, lower(col1_3) ASC, col1_4 ASC, col1_9 DESC, col1_6 ASC) WHERE (table1.col1_9 = 9223372036854775807:::INT8) AND (table1.col1_8 <= (-1):::INT8), FAMILY (col1_9), FAMILY (col1_6, col1_1), FAMILY (col1_5, col1_2, col1_3, col1_0, col1_4));
        ALTER TABLE table1 INJECT STATISTICS e'[{"columns": ["col1_2"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 1341518623222125063, "histo_buckets": [{"distinct_range": 0, "num_eq": 0, "num_range": 0, "upper_bound": "\\r{\\u001b:iX\\u001f"}, {"distinct_range": 0, "num_eq": 7919681357345372707, "num_range": 6775640253247696134, "upper_bound": "/\\u0002\\u0005\\u00198\\u001dVo"}, {"distinct_range": 0, "num_eq": 4000000000, "num_range": 4575478630384794353, "upper_bound": "0u\\u0011t0\\u0018\\u001a"}, {"distinct_range": 0, "num_eq": 8000000, "num_range": 7445320505268038459, "upper_bound": "Fi"}, {"distinct_range": 43215672185694120, "num_eq": 60, "num_range": 2289308350295896476, "upper_bound": "IP[\\u0007;"}, {"distinct_range": 1000000, "num_eq": 4000000, "num_range": 1000000, "upper_bound": "X"}], "histo_col_type": "VARCHAR", "name": "__auto__", "null_count": 0, "row_count": 7985800102032957398}, {"columns": ["col1_4"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 3396083394188221005, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 7985800102032957398}, {"columns": ["col1_8"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 771559412794220387, "histo_buckets": [{"distinct_range": 0, "num_eq": 30000, "num_range": 0, "upper_bound": "-5264861656277209271"}, {"distinct_range": 0, "num_eq": 8021916300917840286, "num_range": 600000000, "upper_bound": "-3940691932668030182"}, {"distinct_range": 1086194854378727600, "num_eq": 1000, "num_range": 1086194854378727501, "upper_bound": "-2925562594724109384"}, {"distinct_range": 2213142197692203300, "num_eq": 7352243173221273865, "num_range": 3395011120584003860, "upper_bound": "1477852551707052052"}], "histo_col_type": "INT8", "name": "__auto__", "null_count": 7459236463088692010, "row_count": 7985800102032957398}, {"columns": ["col1_9"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 4264028756560912785, "histo_buckets": [{"distinct_range": 0, "num_eq": 6000000, "num_range": 0, "upper_bound": "-6466661663957545003"}, {"distinct_range": 4721195992469959000, "num_eq": 0, "num_range": 8549791645122041298, "upper_bound": "-6060510764208708069"}, {"distinct_range": 7943135097669907000, "num_eq": 3000000000, "num_range": 7943135097669907162, "upper_bound": "1"}, {"distinct_range": 100000, "num_eq": 1000000000, "num_range": 100000, "upper_bound": "3348428576045118859"}], "histo_col_type": "INT8", "name": "__auto__", "null_count": 6364933235420285084, "row_count": 7985800102032957398}, {"columns": ["col1_0"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 2734227821202875174, "histo_col_type": "BIT(47)", "name": "__auto__", "null_count": 0, "row_count": 7985800102032957398}, {"columns": ["col1_1"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 505652838481684921, "histo_buckets": [{"distinct_range": 0, "num_eq": 1775048151955946245, "num_range": 0, "upper_bound": "2012-07-26 22:00:24.000111"}], "histo_col_type": "TIMESTAMP", "name": "__auto__", "null_count": 0, "row_count": 7985800102032957398}, {"columns": ["col1_3"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 7611798517405339066, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 7985800102032957398}, {"columns": ["col1_5"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 4533894168597454973, "histo_col_type": "", "name": "__auto__", "null_count": 3515062489097458492, "row_count": 7985800102032957398}, {"columns": ["col1_6"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 1625241631928242707, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 7985800102032957398}, {"columns": ["col1_7"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 5700497799412790850, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 7985800102032957398}]':::JSONB;
        CREATE TYPE rand_typ_0 AS ENUM ('uqt', 'llnjj');
        CREATE TYPE rand_typ_1 AS ENUM ('ki', 'js', 'bn', 'cj', 'lfd');
        CREATE TYPE rand_typ_2 AS ENUM ('z');
        CREATE TYPE rand_typ_3 AS ENUM ('mer', 'qnx', 'ulk', 'yg', 'hpqc');
        CREATE TYPE rand_typ_4 AS ENUM ('ni', 'jqohg');
        ;
    rsg_test.go:598: 
        CREATE TYPE greeting AS ENUM ('hello', 'howdy', 'hi', 'good day', 'morning');
        CREATE TABLE IF NOT EXISTS seed AS
        	SELECT
        		g::INT2 AS _int2,
        		g::INT4 AS _int4,
        		g::INT8 AS _int8,
        		g::FLOAT4 AS _float4,
        		g::FLOAT8 AS _float8,
        		'2001-01-01'::DATE + g AS _date,
        		'2001-01-01'::TIMESTAMP + g * '1 day'::INTERVAL AS _timestamp,
        		'2001-01-01'::TIMESTAMPTZ + g * '1 day'::INTERVAL AS _timestamptz,
        		g * '1 day'::INTERVAL AS _interval,
        		g % 2 = 1 AS _bool,
        		g::DECIMAL AS _decimal,
        		g::STRING AS _string,
        		g::STRING::BYTES AS _bytes,
        		substring('00000000-0000-0000-0000-' || g::STRING || '00000000000', 1, 36)::UUID AS _uuid,
        		'0.0.0.0'::INET + g AS _inet,
        		g::STRING::JSONB AS _jsonb,
        		enum_range('hello'::greeting)[g] as _enum
        	FROM
        		generate_series(1, 5) AS g;
        
        INSERT INTO seed DEFAULT VALUES;
        CREATE INDEX on seed (_int8, _float8, _date);
        CREATE INVERTED INDEX on seed (_jsonb);
        ;
    rsg_test.go:600: 
    rsg_test.go:601: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestRandomSyntaxSQLSmith442408547
--- FAIL: TestRandomSyntaxSQLSmith (366.37s)
```
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestRandomSyntaxSQLSmith PKG=./pkg/sql/tests TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #64592 sql/tests: TestRandomSyntaxSQLSmith failed [C-test-failure O-robot T-sql-queries branch-release-21.1]
</p>
</details>
<details><summary>Internal log</summary>
<p>

```
mjibson marked as alumn{us/a}; resolving to rafiss instead
```
</p>
</details>

/cc @cockroachdb/sql-experience rafiss
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRandomSyntaxSQLSmith.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/sql/logictest/testdata/logic_test/array
+++ b/pkg/sql/logictest/testdata/logic_test/array
@@ -1938,3 +1938,42 @@ query T
 SELECT array_positions(NULL::record[], ROW(33,'hippo'))
 ----
 NULL
+
+# Handle tuples with mismatched types. This behavior matches PostgreSQL -- the
+# tuple comparison will short-circuit, so for example, there will be no error
+# if the second elements' types don't match as long as the first elements are
+# unequal.
+
+query error pgcode 42804 unsupported comparison.*\n.*at record column 2
+SELECT array_replace(ARRAY[(1,'cat'),(2,'dog')], (1, 12.3::decimal), (3,'fish'))
+
+# array_append does not require the tuple types to match exactly.
+query error pgcode 42804 unsupported comparison.*\n.*at record column 2
+SELECT array_replace(array_append(array[(1,'cat')], (2,true)), (2,'fish'), (3,'dog'))
+
+# This works because the comparison short-circuits before comparing the boolean column
+query T
+SELECT array_replace(array_append(array[(1,'cat')], (2,true)), (1,'fish'), (3,'dog'))
+----
+{"(1,cat)","(2,t)"}
+
+# The value that replaces the old one can have types.
+query T
+SELECT array_replace(ARRAY[(1,'cat'),(2,'dog')], (2, 'dog'), (3,true))
+----
+{"(1,cat)","(3,t)"}
+
+query error pgcode 42804 unsupported comparison.*\n.*at record column 1
+SELECT array_position(ARRAY[(12::OID,NULL),(NULL,NULL)], ('\xc03b4478eb'::BYTEA,NULL))
+
+query error pgcode 42804 unsupported comparison.*\n.*at record column 2
+SELECT array_positions(ARRAY[(1,'cat'),(2,'dog')], (2,true))
+
+query error pgcode 42804 unsupported comparison.*\n.*at record column 2
+SELECT array_remove(ARRAY[(1,'cat'),(2,'dog')], (2,'\xc03b4478eb'::BYTEA))
+
+# This works because the comparison short-circuits before comparing the BYTEA column
+query T
+SELECT array_remove(ARRAY[(1,'cat'),(2,'dog')], (3,'\xc03b4478eb'::BYTEA))
+----
+{"(1,cat)","(2,dog)"}

--- a/pkg/sql/sem/builtins/builtins.go
+++ b/pkg/sql/sem/builtins/builtins.go
@@ -3477,7 +3477,11 @@ value if you rely on the HLC for accuracy.`,
 				}
 				result := tree.NewDArray(typ)
 				for _, e := range tree.MustBeDArray(args[0]).Array {
-					if e.Compare(ctx, args[1]) != 0 {
+					cmp, err := e.CompareError(ctx, args[1])
+					if err != nil {
+						return nil, err
+					}
+					if cmp != 0 {
 						if err := result.Append(e); err != nil {
 							return nil, err
 						}
@@ -3500,7 +3504,11 @@ value if you rely on the HLC for accuracy.`,
 				}
 				result := tree.NewDArray(typ)
 				for _, e := range tree.MustBeDArray(args[0]).Array {
-					if e.Compare(ctx, args[1]) == 0 {
+					cmp, err := e.CompareError(ctx, args[1])
+					if err != nil {
+						return nil, err
+					}
+					if cmp == 0 {
 						if err := result.Append(args[2]); err != nil {
 							return nil, err
 						}
@@ -3526,7 +3534,11 @@ value if you rely on the HLC for accuracy.`,
 					return tree.DNull, nil
 				}
 				for i, e := range tree.MustBeDArray(args[0]).Array {
-					if e.Compare(ctx, args[1]) == 0 {
+					cmp, err := e.CompareError(ctx, args[1])
+					if err != nil {
+						return nil, err
+					}
+					if cmp == 0 {
 						return tree.NewDInt(tree.DInt(i + 1)), nil
 					}
 				}
@@ -3547,7 +3559,11 @@ value if you rely on the HLC for accuracy.`,
 				}
 				result := tree.NewDArray(types.Int)
 				for i, e := range tree.MustBeDArray(args[0]).Array {
-					if e.Compare(ctx, args[1]) == 0 {
+					cmp, err := e.CompareError(ctx, args[1])
+					if err != nil {
+						return nil, err
+					}
+					if cmp == 0 {
 						if err := result.Append(tree.NewDInt(tree.DInt(i + 1))); err != nil {
 							return nil, err
 						}

--- a/pkg/sql/sem/tree/datum.go
+++ b/pkg/sql/sem/tree/datum.go
@@ -96,7 +96,12 @@ type Datum interface {
 
 	// Compare returns -1 if the receiver is less than other, 0 if receiver is
 	// equal to other and +1 if receiver is greater than other.
+	// TODO(rafi): Migrate all usages of this to CompareError, then delete this.
 	Compare(ctx *EvalContext, other Datum) int
+
+	// CompareError is the same as Compare, but it returns an error instead of
+	// panicking.
+	CompareError(ctx *EvalContext, other Datum) (int, error)
 
 	// Prev returns the previous datum and true, if one exists, or nil and false.
 	// The previous datum satisfies the following definition: if the receiver is
@@ -407,15 +412,25 @@ func (*DBool) ResolvedType() *types.T {
 
 // Compare implements the Datum interface.
 func (d *DBool) Compare(ctx *EvalContext, other Datum) int {
+	res, err := d.CompareError(ctx, other)
+	if err != nil {
+		panic(err)
+	}
+	return res
+}
+
+// CompareError implements the Datum interface.
+func (d *DBool) CompareError(ctx *EvalContext, other Datum) (int, error) {
 	if other == DNull {
 		// NULL is less than any non-NULL value.
-		return 1
+		return 1, nil
 	}
 	v, ok := UnwrapDatum(ctx, other).(*DBool)
 	if !ok {
-		panic(makeUnsupportedComparisonMessage(d, other))
+		return 0, makeUnsupportedComparisonMessage(d, other)
 	}
-	return CompareBools(bool(*d), bool(*v))
+	res := CompareBools(bool(*d), bool(*v))
+	return res, nil
 }
 
 // CompareBools compares the input bools according to the SQL comparison rules.
@@ -570,15 +585,25 @@ func (*DBitArray) ResolvedType() *types.T {
 
 // Compare implements the Datum interface.
 func (d *DBitArray) Compare(ctx *EvalContext, other Datum) int {
+	res, err := d.CompareError(ctx, other)
+	if err != nil {
+		panic(err)
+	}
+	return res
+}
+
+// CompareError implements the Datum interface.
+func (d *DBitArray) CompareError(ctx *EvalContext, other Datum) (int, error) {
 	if other == DNull {
 		// NULL is less than any non-NULL value.
-		return 1
+		return 1, nil
 	}
 	v, ok := UnwrapDatum(ctx, other).(*DBitArray)
 	if !ok {
-		panic(makeUnsupportedComparisonMessage(d, other))
+		return 0, makeUnsupportedComparisonMessage(d, other)
 	}
-	return bitarray.Compare(d.BitArray, v.BitArray)
+	res := bitarray.Compare(d.BitArray, v.BitArray)
+	return res, nil
 }
 
 // Prev implements the Datum interface.
@@ -688,9 +713,18 @@ func (*DInt) ResolvedType() *types.T {
 
 // Compare implements the Datum interface.
 func (d *DInt) Compare(ctx *EvalContext, other Datum) int {
+	res, err := d.CompareError(ctx, other)
+	if err != nil {
+		panic(err)
+	}
+	return res
+}
+
+// CompareError implements the Datum interface.
+func (d *DInt) CompareError(ctx *EvalContext, other Datum) (int, error) {
 	if other == DNull {
 		// NULL is less than any non-NULL value.
-		return 1
+		return 1, nil
 	}
 	thisInt := *d
 	var v DInt
@@ -698,7 +732,11 @@ func (d *DInt) Compare(ctx *EvalContext, other Datum) int {
 	case *DInt:
 		v = *t
 	case *DFloat, *DDecimal:
-		return -t.Compare(ctx, d)
+		res, err := t.CompareError(ctx, d)
+		if err != nil {
+			return 0, err
+		}
+		return -res, nil
 	case *DOid:
 		// OIDs are always unsigned 32-bit integers. Some languages, like Java,
 		// compare OIDs to signed 32-bit integers, so we implement the comparison
@@ -706,15 +744,15 @@ func (d *DInt) Compare(ctx *EvalContext, other Datum) int {
 		thisInt = DInt(uint32(thisInt))
 		v = t.DInt
 	default:
-		panic(makeUnsupportedComparisonMessage(d, other))
+		return 0, makeUnsupportedComparisonMessage(d, other)
 	}
 	if thisInt < v {
-		return -1
+		return -1, nil
 	}
 	if thisInt > v {
-		return 1
+		return 1, nil
 	}
-	return 0
+	return 0, nil
 }
 
 // Prev implements the Datum interface.
@@ -810,9 +848,18 @@ func (*DFloat) ResolvedType() *types.T {
 
 // Compare implements the Datum interface.
 func (d *DFloat) Compare(ctx *EvalContext, other Datum) int {
+	res, err := d.CompareError(ctx, other)
+	if err != nil {
+		panic(err)
+	}
+	return res
+}
+
+// CompareError implements the Datum interface.
+func (d *DFloat) CompareError(ctx *EvalContext, other Datum) (int, error) {
 	if other == DNull {
 		// NULL is less than any non-NULL value.
-		return 1
+		return 1, nil
 	}
 	var v DFloat
 	switch t := UnwrapDatum(ctx, other).(type) {
@@ -821,27 +868,31 @@ func (d *DFloat) Compare(ctx *EvalContext, other Datum) int {
 	case *DInt:
 		v = DFloat(MustBeDInt(t))
 	case *DDecimal:
-		return -t.Compare(ctx, d)
+		res, err := t.CompareError(ctx, d)
+		if err != nil {
+			return 0, err
+		}
+		return -res, nil
 	default:
-		panic(makeUnsupportedComparisonMessage(d, other))
+		return 0, makeUnsupportedComparisonMessage(d, other)
 	}
 	if *d < v {
-		return -1
+		return -1, nil
 	}
 	if *d > v {
-		return 1
+		return 1, nil
 	}
 	// NaN sorts before non-NaN (#10109).
 	if *d == v {
-		return 0
+		return 0, nil
 	}
 	if math.IsNaN(float64(*d)) {
 		if math.IsNaN(float64(v)) {
-			return 0
+			return 0, nil
 		}
-		return -1
+		return -1, nil
 	}
-	return 1
+	return 1, nil
 }
 
 // Prev implements the Datum interface.
@@ -990,9 +1041,18 @@ func (*DDecimal) ResolvedType() *types.T {
 
 // Compare implements the Datum interface.
 func (d *DDecimal) Compare(ctx *EvalContext, other Datum) int {
+	res, err := d.CompareError(ctx, other)
+	if err != nil {
+		panic(err)
+	}
+	return res
+}
+
+// CompareError implements the Datum interface.
+func (d *DDecimal) CompareError(ctx *EvalContext, other Datum) (int, error) {
 	if other == DNull {
 		// NULL is less than any non-NULL value.
-		return 1
+		return 1, nil
 	}
 	v := ctx.getTmpDec()
 	switch t := UnwrapDatum(ctx, other).(type) {
@@ -1005,9 +1065,10 @@ func (d *DDecimal) Compare(ctx *EvalContext, other Datum) int {
 			panic(errors.NewAssertionErrorWithWrappedErrf(err, "decimal compare, unexpected error"))
 		}
 	default:
-		panic(makeUnsupportedComparisonMessage(d, other))
+		return 0, makeUnsupportedComparisonMessage(d, other)
 	}
-	return CompareDecimals(&d.Decimal, v)
+	res := CompareDecimals(&d.Decimal, v)
+	return res, nil
 }
 
 // CompareDecimals compares 2 apd.Decimals according to the SQL comparison
@@ -1157,21 +1218,30 @@ func (*DString) ResolvedType() *types.T {
 
 // Compare implements the Datum interface.
 func (d *DString) Compare(ctx *EvalContext, other Datum) int {
+	res, err := d.CompareError(ctx, other)
+	if err != nil {
+		panic(err)
+	}
+	return res
+}
+
+// CompareError implements the Datum interface.
+func (d *DString) CompareError(ctx *EvalContext, other Datum) (int, error) {
 	if other == DNull {
 		// NULL is less than any non-NULL value.
-		return 1
+		return 1, nil
 	}
 	v, ok := UnwrapDatum(ctx, other).(*DString)
 	if !ok {
-		panic(makeUnsupportedComparisonMessage(d, other))
+		return 0, makeUnsupportedComparisonMessage(d, other)
 	}
 	if *d < *v {
-		return -1
+		return -1, nil
 	}
 	if *d > *v {
-		return 1
+		return 1, nil
 	}
-	return 0
+	return 0, nil
 }
 
 // Prev implements the Datum interface.
@@ -1303,15 +1373,25 @@ func (d *DCollatedString) ResolvedType() *types.T {
 
 // Compare implements the Datum interface.
 func (d *DCollatedString) Compare(ctx *EvalContext, other Datum) int {
+	res, err := d.CompareError(ctx, other)
+	if err != nil {
+		panic(err)
+	}
+	return res
+}
+
+// CompareError implements the Datum interface.
+func (d *DCollatedString) CompareError(ctx *EvalContext, other Datum) (int, error) {
 	if other == DNull {
 		// NULL is less than any non-NULL value.
-		return 1
+		return 1, nil
 	}
 	v, ok := UnwrapDatum(ctx, other).(*DCollatedString)
 	if !ok || !d.ResolvedType().Equivalent(other.ResolvedType()) {
-		panic(makeUnsupportedComparisonMessage(d, other))
+		return 0, makeUnsupportedComparisonMessage(d, other)
 	}
-	return bytes.Compare(d.Key, v.Key)
+	res := bytes.Compare(d.Key, v.Key)
+	return res, nil
 }
 
 // Prev implements the Datum interface.
@@ -1390,21 +1470,30 @@ func (*DBytes) ResolvedType() *types.T {
 
 // Compare implements the Datum interface.
 func (d *DBytes) Compare(ctx *EvalContext, other Datum) int {
+	res, err := d.CompareError(ctx, other)
+	if err != nil {
+		panic(err)
+	}
+	return res
+}
+
+// CompareError implements the Datum interface.
+func (d *DBytes) CompareError(ctx *EvalContext, other Datum) (int, error) {
 	if other == DNull {
 		// NULL is less than any non-NULL value.
-		return 1
+		return 1, nil
 	}
 	v, ok := UnwrapDatum(ctx, other).(*DBytes)
 	if !ok {
-		panic(makeUnsupportedComparisonMessage(d, other))
+		return 0, makeUnsupportedComparisonMessage(d, other)
 	}
 	if *d < *v {
-		return -1
+		return -1, nil
 	}
 	if *d > *v {
-		return 1
+		return 1, nil
 	}
-	return 0
+	return 0, nil
 }
 
 // Prev implements the Datum interface.
@@ -1497,15 +1586,25 @@ func (*DUuid) ResolvedType() *types.T {
 
 // Compare implements the Datum interface.
 func (d *DUuid) Compare(ctx *EvalContext, other Datum) int {
+	res, err := d.CompareError(ctx, other)
+	if err != nil {
+		panic(err)
+	}
+	return res
+}
+
+// CompareError implements the Datum interface.
+func (d *DUuid) CompareError(ctx *EvalContext, other Datum) (int, error) {
 	if other == DNull {
 		// NULL is less than any non-NULL value.
-		return 1
+		return 1, nil
 	}
 	v, ok := UnwrapDatum(ctx, other).(*DUuid)
 	if !ok {
-		panic(makeUnsupportedComparisonMessage(d, other))
+		return 0, makeUnsupportedComparisonMessage(d, other)
 	}
-	return bytes.Compare(d.GetBytes(), v.GetBytes())
+	res := bytes.Compare(d.GetBytes(), v.GetBytes())
+	return res, nil
 }
 
 func (d *DUuid) equal(other *DUuid) bool {
@@ -1616,16 +1715,26 @@ func (*DIPAddr) ResolvedType() *types.T {
 
 // Compare implements the Datum interface.
 func (d *DIPAddr) Compare(ctx *EvalContext, other Datum) int {
+	res, err := d.CompareError(ctx, other)
+	if err != nil {
+		panic(err)
+	}
+	return res
+}
+
+// CompareError implements the Datum interface.
+func (d *DIPAddr) CompareError(ctx *EvalContext, other Datum) (int, error) {
 	if other == DNull {
 		// NULL is less than any non-NULL value.
-		return 1
+		return 1, nil
 	}
 	v, ok := UnwrapDatum(ctx, other).(*DIPAddr)
 	if !ok {
-		panic(makeUnsupportedComparisonMessage(d, other))
+		return 0, makeUnsupportedComparisonMessage(d, other)
 	}
 
-	return d.IPAddr.Compare(&v.IPAddr)
+	res := d.IPAddr.Compare(&v.IPAddr)
+	return res, nil
 }
 
 func (d DIPAddr) equal(other *DIPAddr) bool {
@@ -1888,9 +1997,18 @@ func (*DDate) ResolvedType() *types.T {
 
 // Compare implements the Datum interface.
 func (d *DDate) Compare(ctx *EvalContext, other Datum) int {
+	res, err := d.CompareError(ctx, other)
+	if err != nil {
+		panic(err)
+	}
+	return res
+}
+
+// CompareError implements the Datum interface.
+func (d *DDate) CompareError(ctx *EvalContext, other Datum) (int, error) {
 	if other == DNull {
 		// NULL is less than any non-NULL value.
-		return 1
+		return 1, nil
 	}
 	var v DDate
 	switch t := UnwrapDatum(ctx, other).(type) {
@@ -1899,9 +2017,10 @@ func (d *DDate) Compare(ctx *EvalContext, other Datum) int {
 	case *DTimestamp, *DTimestampTZ:
 		return compareTimestamps(ctx, d, other)
 	default:
-		panic(makeUnsupportedComparisonMessage(d, other))
+		return 0, makeUnsupportedComparisonMessage(d, other)
 	}
-	return d.Date.Compare(v.Date)
+	res := d.Date.Compare(v.Date)
+	return res, nil
 }
 
 var (
@@ -2035,9 +2154,18 @@ func (*DTime) ResolvedType() *types.T {
 
 // Compare implements the Datum interface.
 func (d *DTime) Compare(ctx *EvalContext, other Datum) int {
+	res, err := d.CompareError(ctx, other)
+	if err != nil {
+		panic(err)
+	}
+	return res
+}
+
+// CompareError implements the Datum interface.
+func (d *DTime) CompareError(ctx *EvalContext, other Datum) (int, error) {
 	if other == DNull {
 		// NULL is less than any non-NULL value.
-		return 1
+		return 1, nil
 	}
 	return compareTimestamps(ctx, d, other)
 }
@@ -2165,9 +2293,18 @@ func (*DTimeTZ) ResolvedType() *types.T {
 
 // Compare implements the Datum interface.
 func (d *DTimeTZ) Compare(ctx *EvalContext, other Datum) int {
+	res, err := d.CompareError(ctx, other)
+	if err != nil {
+		panic(err)
+	}
+	return res
+}
+
+// CompareError implements the Datum interface.
+func (d *DTimeTZ) CompareError(ctx *EvalContext, other Datum) (int, error) {
 	if other == DNull {
 		// NULL is less than any non-NULL value.
-		return 1
+		return 1, nil
 	}
 	return compareTimestamps(ctx, d, other)
 }
@@ -2380,7 +2517,7 @@ func checkInfiniteDate(ctx *EvalContext, d Datum) infiniteDateComparison {
 // Datums are allowed to be one of DDate, DTimestamp, DTimestampTZ, DTime,
 // DTimeTZ. For all other datum types it will panic; also, comparing two DDates
 // is not supported.
-func compareTimestamps(ctx *EvalContext, l Datum, r Datum) int {
+func compareTimestamps(ctx *EvalContext, l Datum, r Datum) (int, error) {
 	leftInf := checkInfiniteDate(ctx, l)
 	rightInf := checkInfiniteDate(ctx, r)
 	if leftInf != finite || rightInf != finite {
@@ -2388,23 +2525,23 @@ func compareTimestamps(ctx *EvalContext, l Datum, r Datum) int {
 		if leftInf != finite && rightInf != finite {
 			// Both datums cannot be infinite dates at the same time because we
 			// wouldn't use this method.
-			panic(errors.AssertionFailedf("unexpectedly two infinite dates in compareTimestamps"))
+			return 0, errors.AssertionFailedf("unexpectedly two infinite dates in compareTimestamps")
 		}
 		// Exactly one of the datums is an infinite date and another is a finite
 		// datums (not necessarily a date). We can just subtract the returned
 		// values to get the desired result for comparison.
-		return int(leftInf - rightInf)
+		return int(leftInf - rightInf), nil
 	}
 	lTime, lErr := timeFromDatumForComparison(ctx, l)
 	rTime, rErr := timeFromDatumForComparison(ctx, r)
 	if lErr != nil || rErr != nil {
-		panic(makeUnsupportedComparisonMessage(l, r))
+		return 0, makeUnsupportedComparisonMessage(l, r)
 	}
 	if lTime.Before(rTime) {
-		return -1
+		return -1, nil
 	}
 	if rTime.Before(lTime) {
-		return 1
+		return 1, nil
 	}
 
 	// If either side is a TimeTZ, then we must compare timezones before
@@ -2419,7 +2556,7 @@ func compareTimestamps(ctx *EvalContext, l Datum, r Datum) int {
 
 	// If neither side is TimeTZ, this is always equal at this point.
 	if !leftIsTimeTZ && !rightIsTimeTZ {
-		return 0
+		return 0, nil
 	}
 
 	_, zoneOffset := ctx.GetRelativeParseTime().Zone()
@@ -2434,19 +2571,28 @@ func compareTimestamps(ctx *EvalContext, l Datum, r Datum) int {
 	}
 
 	if lOffset > rOffset {
-		return 1
+		return 1, nil
 	}
 	if lOffset < rOffset {
-		return -1
+		return -1, nil
 	}
-	return 0
+	return 0, nil
 }
 
 // Compare implements the Datum interface.
 func (d *DTimestamp) Compare(ctx *EvalContext, other Datum) int {
+	res, err := d.CompareError(ctx, other)
+	if err != nil {
+		panic(err)
+	}
+	return res
+}
+
+// CompareError implements the Datum interface.
+func (d *DTimestamp) CompareError(ctx *EvalContext, other Datum) (int, error) {
 	if other == DNull {
 		// NULL is less than any non-NULL value.
-		return 1
+		return 1, nil
 	}
 	return compareTimestamps(ctx, d, other)
 }
@@ -2601,9 +2747,18 @@ func (*DTimestampTZ) ResolvedType() *types.T {
 
 // Compare implements the Datum interface.
 func (d *DTimestampTZ) Compare(ctx *EvalContext, other Datum) int {
+	res, err := d.CompareError(ctx, other)
+	if err != nil {
+		panic(err)
+	}
+	return res
+}
+
+// CompareError implements the Datum interface.
+func (d *DTimestampTZ) CompareError(ctx *EvalContext, other Datum) (int, error) {
 	if other == DNull {
 		// NULL is less than any non-NULL value.
-		return 1
+		return 1, nil
 	}
 	return compareTimestamps(ctx, d, other)
 }
@@ -2808,15 +2963,25 @@ func (*DInterval) ResolvedType() *types.T {
 
 // Compare implements the Datum interface.
 func (d *DInterval) Compare(ctx *EvalContext, other Datum) int {
+	res, err := d.CompareError(ctx, other)
+	if err != nil {
+		panic(err)
+	}
+	return res
+}
+
+// CompareError implements the Datum interface.
+func (d *DInterval) CompareError(ctx *EvalContext, other Datum) (int, error) {
 	if other == DNull {
 		// NULL is less than any non-NULL value.
-		return 1
+		return 1, nil
 	}
 	v, ok := UnwrapDatum(ctx, other).(*DInterval)
 	if !ok {
-		panic(makeUnsupportedComparisonMessage(d, other))
+		return 0, makeUnsupportedComparisonMessage(d, other)
 	}
-	return d.Duration.Compare(v.Duration)
+	res := d.Duration.Compare(v.Duration)
+	return res, nil
 }
 
 // Prev implements the Datum interface.
@@ -2936,11 +3101,25 @@ func (*DGeography) ResolvedType() *types.T {
 
 // Compare implements the Datum interface.
 func (d *DGeography) Compare(ctx *EvalContext, other Datum) int {
+	res, err := d.CompareError(ctx, other)
+	if err != nil {
+		panic(err)
+	}
+	return res
+}
+
+// CompareError implements the Datum interface.
+func (d *DGeography) CompareError(ctx *EvalContext, other Datum) (int, error) {
 	if other == DNull {
 		// NULL is less than any non-NULL value.
-		return 1
+		return 1, nil
 	}
-	return d.Geography.Compare(other.(*DGeography).Geography)
+	v, ok := UnwrapDatum(ctx, other).(*DGeography)
+	if !ok {
+		return 0, makeUnsupportedComparisonMessage(d, other)
+	}
+	res := d.Geography.Compare(v.Geography)
+	return res, nil
 }
 
 // Prev implements the Datum interface.
@@ -3044,11 +3223,25 @@ func (*DGeometry) ResolvedType() *types.T {
 
 // Compare implements the Datum interface.
 func (d *DGeometry) Compare(ctx *EvalContext, other Datum) int {
+	res, err := d.CompareError(ctx, other)
+	if err != nil {
+		panic(err)
+	}
+	return res
+}
+
+// CompareError implements the Datum interface.
+func (d *DGeometry) CompareError(ctx *EvalContext, other Datum) (int, error) {
 	if other == DNull {
 		// NULL is less than any non-NULL value.
-		return 1
+		return 1, nil
 	}
-	return d.Geometry.Compare(other.(*DGeometry).Geometry)
+	v, ok := UnwrapDatum(ctx, other).(*DGeometry)
+	if !ok {
+		return 0, makeUnsupportedComparisonMessage(d, other)
+	}
+	res := d.Geometry.Compare(v.Geometry)
+	return res, nil
 }
 
 // Prev implements the Datum interface.
@@ -3152,12 +3345,25 @@ func (*DBox2D) ResolvedType() *types.T {
 
 // Compare implements the Datum interface.
 func (d *DBox2D) Compare(ctx *EvalContext, other Datum) int {
+	res, err := d.CompareError(ctx, other)
+	if err != nil {
+		panic(err)
+	}
+	return res
+}
+
+// CompareError implements the Datum interface.
+func (d *DBox2D) CompareError(ctx *EvalContext, other Datum) (int, error) {
 	if other == DNull {
 		// NULL is less than any non-NULL value.
-		return 1
+		return 1, nil
 	}
-	o := other.(*DBox2D)
-	return d.CartesianBoundingBox.Compare(&o.CartesianBoundingBox)
+	v, ok := UnwrapDatum(ctx, other).(*DBox2D)
+	if !ok {
+		return 0, makeUnsupportedComparisonMessage(d, other)
+	}
+	res := d.CartesianBoundingBox.Compare(&v.CartesianBoundingBox)
+	return res, nil
 }
 
 // Prev implements the Datum interface.
@@ -3354,22 +3560,28 @@ func (*DJSON) ResolvedType() *types.T {
 
 // Compare implements the Datum interface.
 func (d *DJSON) Compare(ctx *EvalContext, other Datum) int {
-	if other == DNull {
-		// NULL is less than any non-NULL value.
-		return 1
-	}
-	v, ok := UnwrapDatum(ctx, other).(*DJSON)
-	if !ok {
-		panic(makeUnsupportedComparisonMessage(d, other))
-	}
-	// No avenue for us to pass up this error here at the moment, but Compare
-	// only errors for invalid encoded data.
-	// TODO(justin): modify Compare to allow passing up errors.
-	c, err := d.JSON.Compare(v.JSON)
+	res, err := d.CompareError(ctx, other)
 	if err != nil {
 		panic(err)
 	}
-	return c
+	return res
+}
+
+// CompareError implements the Datum interface.
+func (d *DJSON) CompareError(ctx *EvalContext, other Datum) (int, error) {
+	if other == DNull {
+		// NULL is less than any non-NULL value.
+		return 1, nil
+	}
+	v, ok := UnwrapDatum(ctx, other).(*DJSON)
+	if !ok {
+		return 0, makeUnsupportedComparisonMessage(d, other)
+	}
+	c, err := d.JSON.Compare(v.JSON)
+	if err != nil {
+		return 0, err
+	}
+	return c, nil
 }
 
 // Prev implements the Datum interface.
@@ -3500,31 +3712,43 @@ func (d *DTuple) ResolvedType() *types.T {
 
 // Compare implements the Datum interface.
 func (d *DTuple) Compare(ctx *EvalContext, other Datum) int {
+	res, err := d.CompareError(ctx, other)
+	if err != nil {
+		panic(err)
+	}
+	return res
+}
+
+// CompareError implements the Datum interface.
+func (d *DTuple) CompareError(ctx *EvalContext, other Datum) (int, error) {
 	if other == DNull {
 		// NULL is less than any non-NULL value.
-		return 1
+		return 1, nil
 	}
 	v, ok := UnwrapDatum(ctx, other).(*DTuple)
 	if !ok {
-		panic(makeUnsupportedComparisonMessage(d, other))
+		return 0, makeUnsupportedComparisonMessage(d, other)
 	}
 	n := len(d.D)
 	if n > len(v.D) {
 		n = len(v.D)
 	}
 	for i := 0; i < n; i++ {
-		c := d.D[i].Compare(ctx, v.D[i])
+		c, err := d.D[i].CompareError(ctx, v.D[i])
+		if err != nil {
+			return 0, err
+		}
 		if c != 0 {
-			return c
+			return c, nil
 		}
 	}
 	if len(d.D) < len(v.D) {
-		return -1
+		return -1, nil
 	}
 	if len(d.D) > len(v.D) {
-		return 1
+		return 1, nil
 	}
-	return 0
+	return 0, nil
 }
 
 // Prev implements the Datum interface.
@@ -3807,10 +4031,19 @@ func (dNull) ResolvedType() *types.T {
 
 // Compare implements the Datum interface.
 func (d dNull) Compare(ctx *EvalContext, other Datum) int {
-	if other == DNull {
-		return 0
+	res, err := d.CompareError(ctx, other)
+	if err != nil {
+		panic(err)
 	}
-	return -1
+	return res
+}
+
+// CompareError implements the Datum interface.
+func (d dNull) CompareError(ctx *EvalContext, other Datum) (int, error) {
+	if other == DNull {
+		return 0, nil
+	}
+	return -1, nil
 }
 
 // Prev implements the Datum interface.
@@ -3940,31 +4173,43 @@ func (d *DArray) FirstIndex() int {
 
 // Compare implements the Datum interface.
 func (d *DArray) Compare(ctx *EvalContext, other Datum) int {
+	res, err := d.CompareError(ctx, other)
+	if err != nil {
+		panic(err)
+	}
+	return res
+}
+
+// CompareError implements the Datum interface.
+func (d *DArray) CompareError(ctx *EvalContext, other Datum) (int, error) {
 	if other == DNull {
 		// NULL is less than any non-NULL value.
-		return 1
+		return 1, nil
 	}
 	v, ok := UnwrapDatum(ctx, other).(*DArray)
 	if !ok {
-		panic(makeUnsupportedComparisonMessage(d, other))
+		return 0, makeUnsupportedComparisonMessage(d, other)
 	}
 	n := d.Len()
 	if n > v.Len() {
 		n = v.Len()
 	}
 	for i := 0; i < n; i++ {
-		c := d.Array[i].Compare(ctx, v.Array[i])
+		c, err := d.Array[i].CompareError(ctx, v.Array[i])
+		if err != nil {
+			return 0, err
+		}
 		if c != 0 {
-			return c
+			return c, nil
 		}
 	}
 	if d.Len() < v.Len() {
-		return -1
+		return -1, nil
 	}
 	if d.Len() > v.Len() {
-		return 1
+		return 1, nil
 	}
-	return 0
+	return 0, nil
 }
 
 // Prev implements the Datum interface.
@@ -4227,14 +4472,24 @@ func (d *DEnum) ResolvedType() *types.T {
 
 // Compare implements the Datum interface.
 func (d *DEnum) Compare(ctx *EvalContext, other Datum) int {
+	res, err := d.CompareError(ctx, other)
+	if err != nil {
+		panic(err)
+	}
+	return res
+}
+
+// CompareError implements the Datum interface.
+func (d *DEnum) CompareError(ctx *EvalContext, other Datum) (int, error) {
 	if other == DNull {
-		return 1
+		return 1, nil
 	}
 	v, ok := UnwrapDatum(ctx, other).(*DEnum)
 	if !ok {
-		panic(makeUnsupportedComparisonMessage(d, other))
+		return 0, makeUnsupportedComparisonMessage(d, other)
 	}
-	return bytes.Compare(d.PhysicalRep, v.PhysicalRep)
+	res := bytes.Compare(d.PhysicalRep, v.PhysicalRep)
+	return res, nil
 }
 
 // Prev implements the Datum interface.
@@ -4663,9 +4918,18 @@ func (*DOid) AmbiguousFormat() bool { return true }
 
 // Compare implements the Datum interface.
 func (d *DOid) Compare(ctx *EvalContext, other Datum) int {
+	res, err := d.CompareError(ctx, other)
+	if err != nil {
+		panic(err)
+	}
+	return res
+}
+
+// CompareError implements the Datum interface.
+func (d *DOid) CompareError(ctx *EvalContext, other Datum) (int, error) {
 	if other == DNull {
 		// NULL is less than any non-NULL value.
-		return 1
+		return 1, nil
 	}
 	var v DInt
 	switch t := UnwrapDatum(ctx, other).(type) {
@@ -4677,16 +4941,16 @@ func (d *DOid) Compare(ctx *EvalContext, other Datum) int {
 		// by converting to a uint32 first. This matches Postgres behavior.
 		v = DInt(uint32(*t))
 	default:
-		panic(makeUnsupportedComparisonMessage(d, other))
+		return 0, makeUnsupportedComparisonMessage(d, other)
 	}
 
 	if d.DInt < v {
-		return -1
+		return -1, nil
 	}
 	if d.DInt > v {
-		return 1
+		return 1, nil
 	}
-	return 0
+	return 0, nil
 }
 
 // Format implements the Datum interface.
@@ -4823,14 +5087,23 @@ func (d *DOidWrapper) ResolvedType() *types.T {
 
 // Compare implements the Datum interface.
 func (d *DOidWrapper) Compare(ctx *EvalContext, other Datum) int {
+	res, err := d.CompareError(ctx, other)
+	if err != nil {
+		panic(err)
+	}
+	return res
+}
+
+// CompareError implements the Datum interface.
+func (d *DOidWrapper) CompareError(ctx *EvalContext, other Datum) (int, error) {
 	if other == DNull {
 		// NULL is less than any non-NULL value.
-		return 1
+		return 1, nil
 	}
 	if v, ok := other.(*DOidWrapper); ok {
-		return d.Wrapped.Compare(ctx, v.Wrapped)
+		return d.Wrapped.CompareError(ctx, v.Wrapped)
 	}
-	return d.Wrapped.Compare(ctx, other)
+	return d.Wrapped.CompareError(ctx, other)
 }
 
 // Prev implements the Datum interface.
@@ -4902,7 +5175,16 @@ func (d *Placeholder) mustGetValue(ctx *EvalContext) Datum {
 
 // Compare implements the Datum interface.
 func (d *Placeholder) Compare(ctx *EvalContext, other Datum) int {
-	return d.mustGetValue(ctx).Compare(ctx, other)
+	res, err := d.CompareError(ctx, other)
+	if err != nil {
+		panic(err)
+	}
+	return res
+}
+
+// CompareError implements the Datum interface.
+func (d *Placeholder) CompareError(ctx *EvalContext, other Datum) (int, error) {
+	return d.mustGetValue(ctx).CompareError(ctx, other)
 }
 
 // Prev implements the Datum interface.

--- a/pkg/sql/sem/tree/datum.go
+++ b/pkg/sql/sem/tree/datum.go
@@ -45,6 +45,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/uint128"
 	"github.com/cockroachdb/cockroach/pkg/util/uuid"
 	"github.com/cockroachdb/errors"
+	"github.com/cockroachdb/redact"
 	"github.com/lib/pq/oid"
 	"golang.org/x/text/collate"
 	"golang.org/x/text/language"
@@ -279,8 +280,11 @@ func MakeParseError(s string, typ *types.T, err error) error {
 }
 
 func makeUnsupportedComparisonMessage(d1, d2 Datum) error {
-	return errors.AssertionFailedWithDepthf(1,
-		"unsupported comparison: %s to %s", errors.Safe(d1.ResolvedType()), errors.Safe(d2.ResolvedType()))
+	return pgerror.Newf(pgcode.DatatypeMismatch,
+		"unsupported comparison: %s to %s",
+		errors.Safe(d1.ResolvedType()),
+		errors.Safe(d2.ResolvedType()),
+	)
 }
 
 func isCaseInsensitivePrefix(prefix, s string) bool {
@@ -3736,7 +3740,7 @@ func (d *DTuple) CompareError(ctx *EvalContext, other Datum) (int, error) {
 	for i := 0; i < n; i++ {
 		c, err := d.D[i].CompareError(ctx, v.D[i])
 		if err != nil {
-			return 0, err
+			return 0, errors.WithDetailf(err, "type mismatch at record column %d", redact.SafeInt(i+1))
 		}
 		if c != 0 {
 			return c, nil

--- a/pkg/sql/sem/tree/datum_test.go
+++ b/pkg/sql/sem/tree/datum_test.go
@@ -146,13 +146,17 @@ func TestCompareTimestamps(t *testing.T) {
 						},
 					),
 				}
-				assert.Equal(t, tc.expected, compareTimestamps(ctx, tc.left, tc.right))
-				assert.Equal(t, -tc.expected, compareTimestamps(ctx, tc.right, tc.left))
+				res, err := compareTimestamps(ctx, tc.left, tc.right)
+				assert.NoError(t, err)
+				assert.Equal(t, tc.expected, res)
+				res, err = compareTimestamps(ctx, tc.right, tc.left)
+				assert.NoError(t, err)
+				assert.Equal(t, -tc.expected, res)
 			},
 		)
 	}
-
-	assert.Panics(t, func() { compareTimestamps(nil /* ctx */, dMaxDate, dMinDate) })
+	_, err = compareTimestamps(nil /* ctx */, dMaxDate, dMinDate)
+	assert.Error(t, err, "should not be able to compare infinite timestamps")
 }
 
 func TestCastStringToRegClassTableName(t *testing.T) {

